**Submitted by: Group G4**->(Umesha B K,Raja Ram Chaudhary,Gowthami R)<br>
**2020-06-30**

![](airindia.png)
<h1 style="text-align:center;"> Air Fare prediction Analysis</h1>

 Import necessaries library

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

Read given data by help of pandas function 

In [2]:
data = pd.read_csv('C:\\Users\\Vinayaka\\Desktop\\project2020\\air_fair.csv')

## Exploratory Data Analysis (EDA)<a id="3"></a>

Show top 5 records

In [3]:
data.head()

InvoiceDate  NetFare    ProductType ItineraryType
0     1/12/18 11:43  2160.00  Other Product      Domestic
1     1/12/18 11:43     0.49  Other Product      Domestic
2     1/12/18 13:01  6550.00            Air      Domestic
3     1/12/18 13:10  6550.00            Air      Domestic
4  30/11/2018 14:30  5497.00            Air      Domestic

InvoiceDate converted int datetime format

In [4]:
data['InvoiceDate']=pd.to_datetime(data['InvoiceDate'])

Extraction of Date, Year, Month, Day, Hour & Minute from InvoiceDate columns

In [5]:
data['Date']=data['InvoiceDate'].dt.date
data['Year']=data['InvoiceDate'].dt.year
data['Month']=data['InvoiceDate'].dt.month
data['Day']=data['InvoiceDate'].dt.day
data['Hour']=data['InvoiceDate'].dt.hour
data['Minute']=data['InvoiceDate'].dt.minute
data

InvoiceDate  NetFare    ProductType ItineraryType        Date  \
0      2018-01-12 11:43:00  2160.00  Other Product      Domestic  2018-01-12   
1      2018-01-12 11:43:00     0.49  Other Product      Domestic  2018-01-12   
2      2018-01-12 13:01:00  6550.00            Air      Domestic  2018-01-12   
3      2018-01-12 13:10:00  6550.00            Air      Domestic  2018-01-12   
4      2018-11-30 14:30:00  5497.00            Air      Domestic  2018-11-30   
...                    ...      ...            ...           ...         ...   
278461 2019-10-06 20:45:00     0.00            Air      Domestic  2019-10-06   
278462 2019-10-06 21:39:00  9073.00            Air      Domestic  2019-10-06   
278463 2019-10-06 22:18:00  2367.00            Air      Domestic  2019-10-06   
278464 2019-10-06 22:30:00  6572.00            Air      Domestic  2019-10-06   
278465 2019-10-06 22:30:00  5118.00            Air      Domestic  2019-10-06   

          Year  Month   Day  Hour  Minute  
0       2018.0    1.0  12.0  11.0    43.0  
1       2018.0    1.0  12.0  11.0    43.0  
2       2018.0    1.0  12.0  13.0     1.0  
3       2018.0    1.0  12.0  13.0    10.0  
4       2018.0   11.0  30.0  14.0    30.0  
...        ...    ...   ...   ...     ...  
278461  2019.0   10.0   6.0  20.0    45.0  
278462  2019.0   10.0   6.0  21.0    39.0  
278463  2019.0   10.0   6.0  22.0    18.0  
278464  2019.0   10.0   6.0  22.0    30.0  
278465  2019.0   10.0   6.0  22.0    30.0  

[278466 rows x 10 columns]

Drop InvoiceDate column after formation of Date, Year, Month, Day, Date, Hour & Minute from InvoiceDate columns

In [6]:
data.drop(["InvoiceDate"], axis=1, inplace=True)

Rearrange of columns as our choices

In [7]:
data = data[['Date','Year', 'Month','Day','Hour','Minute','ProductType','ItineraryType','NetFare']]
data.head()

Date    Year  Month   Day  Hour  Minute    ProductType ItineraryType  \
0  2018-01-12  2018.0    1.0  12.0  11.0    43.0  Other Product      Domestic   
1  2018-01-12  2018.0    1.0  12.0  11.0    43.0  Other Product      Domestic   
2  2018-01-12  2018.0    1.0  12.0  13.0     1.0            Air      Domestic   
3  2018-01-12  2018.0    1.0  12.0  13.0    10.0            Air      Domestic   
4  2018-11-30  2018.0   11.0  30.0  14.0    30.0            Air      Domestic   

   NetFare  
0  2160.00  
1     0.49  
2  6550.00  
3  6550.00  
4  5497.00

Converting data types on suitable format as mention below

In [8]:
#data['Year']=pd.to_datetime(data['float'])
data['Year']=data['Year'].astype('float')
data['Month']=data['Month'].astype('float')
data['Day']=data['Day'].astype('float')
data['Hour']=data['Hour'].astype('float')
data['Minute']=data['Minute'].astype('float')

Parse strings to datetime type 

In [9]:
data['Date'] = pd.to_datetime(data.Date, infer_datetime_format=True)
df=data.set_index(['Date'])

Data index sorted

In [10]:
df.sort_values(by='Date').head()

Year  Month  Day  Hour  Minute ProductType ItineraryType  \
Date                                                                     
2018-01-04  2018.0    1.0  4.0   8.0    26.0         Air      Domestic   
2018-01-04  2018.0    1.0  4.0   9.0    17.0         Air      Domestic   
2018-01-04  2018.0    1.0  4.0  22.0    38.0         Air      Domestic   
2018-01-04  2018.0    1.0  4.0  21.0    12.0         Air      Domestic   
2018-01-04  2018.0    1.0  4.0  20.0    59.0         Air      Domestic   

            NetFare  
Date                 
2018-01-04   8222.0  
2018-01-04   3418.0  
2018-01-04   6261.0  
2018-01-04   3473.0  
2018-01-04   5011.0

## Data Preparation <a id="4"></a>

<h4>Evaluating for Missing Data</h4>
There are two methods to detect missing data:
<ol>
 <li><b>.isnull()</b></li>
 <li><b>.notnull()</b></li>
</ol>   

Find missing value by Boolean method:
**"True"** stands for missing value, while **"False"** stands for not missing value.

In [11]:
missing_value = df.isnull()
missing_value

Year  Month    Day   Hour  Minute  ProductType  ItineraryType  \
Date                                                                         
2018-01-12  False  False  False  False   False        False          False   
2018-01-12  False  False  False  False   False        False          False   
2018-01-12  False  False  False  False   False        False          False   
2018-01-12  False  False  False  False   False        False          False   
2018-11-30  False  False  False  False   False        False          False   
...           ...    ...    ...    ...     ...          ...            ...   
2019-10-06  False  False  False  False   False        False          False   
2019-10-06  False  False  False  False   False        False          False   
2019-10-06  False  False  False  False   False        False          False   
2019-10-06  False  False  False  False   False        False          False   
2019-10-06  False  False  False  False   False        False          False   

            NetFare  
Date                 
2018-01-12    False  
2018-01-12    False  
2018-01-12    False  
2018-01-12    False  
2018-11-30    False  
...             ...  
2019-10-06    False  
2019-10-06    False  
2019-10-06    False  
2019-10-06    False  
2019-10-06    False  

[278466 rows x 8 columns]

### Count missing values in each column

Using a for loop in Python, we can quickly figure out the number of missing values in each column.<br>
**"True" stands for missing value, while "False" stands for not missing value**

In [12]:
for column in missing_value.columns.values.tolist():
    print(column)
    print (missing_value[column].value_counts())
    print("") 

Year
False    278464
True          2
Name: Year, dtype: int64

Month
False    278464
True          2
Name: Month, dtype: int64

Day
False    278464
True          2
Name: Day, dtype: int64

Hour
False    278464
True          2
Name: Hour, dtype: int64

Minute
False    278464
True          2
Name: Minute, dtype: int64

ProductType
False    278464
True          2
Name: ProductType, dtype: int64

ItineraryType
False    245689
True      32777
Name: ItineraryType, dtype: int64

NetFare
False    217576
True      60890
Name: NetFare, dtype: int64



Based on the summary above, each column has 278466 rows of data, all columns containing missing data:
<ol>
<li>"Year","Month","Day","Hour","Minute": 2 missing data</li>
<li>"ProductType": 2 missing data</li>
<li>"ItineraryType": 32777 missing data</li>    
<li>"NetFare": 60890 missing data</li>
</ol>

<h3 id="deal_missing_values">Deal with missing data</h3>

<b>Drop rows:</b>
<ol>
    <li>"Year","Month","Day","Hour","Minute": 2 missing data, drop them , two observations among large dataset are not so big deals</li>
    <li>"ProductType": 2 missing data, drop them , two observations among large dataset are not so big deals</li>
    <li>"ItineraryType": 32777 missing data, drop them , they are totally empty, we have huge data</li>
    <li>"NetFare": 60890 missing data, drop whole rows, they are totally empty</li>  
</ol>

Show data types

Drop missing values **["Year","Month","Day","Hour","Minute"]** in observations, the two rows are negelegible

In [13]:
df.dropna(subset=["Year","Month","Day","Hour","Minute"], axis=0, inplace=True)

Drop missing values **[ProductType]** in observations, the two rows are negelegible 

In [14]:
df.dropna(subset=["ProductType"], axis=0, inplace=True)

Drop missing values **[ItineraryType]** in observations, they are not very sensitive 

In [15]:
df.dropna(subset=["ItineraryType"], axis=0, inplace=True)

Drop missing values **[NetFare]** in observations, they are completly empty 

In [16]:
df.dropna(subset=["NetFare"], axis=0, inplace=True)

Reset index, because we droped many rows & sorted

In [17]:
df.reset_index().sort_values(by='Date').head()

Date    Year  Month  Day  Hour  Minute ProductType ItineraryType  \
93239 2018-01-04  2018.0    1.0  4.0  17.0    28.0         Air      Domestic   
93227 2018-01-04  2018.0    1.0  4.0  13.0    25.0         Air      Domestic   
93250 2018-01-04  2018.0    1.0  4.0  22.0    38.0         Air      Domestic   
93249 2018-01-04  2018.0    1.0  4.0  21.0    12.0         Air      Domestic   
93248 2018-01-04  2018.0    1.0  4.0  20.0    59.0         Air      Domestic   

       NetFare  
93239   7515.0  
93227   5168.0  
93250   6261.0  
93249   3473.0  
93248   5011.0

**Final Clean** data

In [18]:
df_clean = df.isnull()

**"True" stands for missing value, while "False" stands for not missing value**

In [19]:
for column in df_clean.columns.values.tolist():
    print(column)
    print (df_clean[column].value_counts())
    print("") 

Year
False    184800
Name: Year, dtype: int64

Month
False    184800
Name: Month, dtype: int64

Day
False    184800
Name: Day, dtype: int64

Hour
False    184800
Name: Hour, dtype: int64

Minute
False    184800
Name: Minute, dtype: int64

ProductType
False    184800
Name: ProductType, dtype: int64

ItineraryType
False    184800
Name: ItineraryType, dtype: int64

NetFare
False    184800
Name: NetFare, dtype: int64



RAW Data observations=**278466**, CLEANED Data observations=**184800**

#### Identifying Outliers with Interquartile Range (IQR)

In [20]:
Q1 = df.quantile(0.25)
Q3 = df.quantile(0.75)
IQR = Q3 - Q1  
print(IQR)

Year          1.0
Month         6.0
Day          18.0
Hour          6.0
Minute       30.0
NetFare    4314.0
dtype: float64


The code below generates an output with the 'True' and 'False' values. Points where the values are 'True' represent the presence of the outlier.

In [21]:
print((df  < (Q1 - 1.5 * IQR)) | (df  > (Q3 + 1.5 * IQR)))

              Day   Hour  ItineraryType  Minute  Month  NetFare  ProductType  \
Date                                                                           
2018-01-12  False  False          False   False  False    False        False   
2018-01-12  False  False          False   False  False    False        False   
2018-01-12  False  False          False   False  False    False        False   
2018-01-12  False  False          False   False  False    False        False   
2018-11-30  False  False          False   False  False    False        False   
...           ...    ...            ...     ...    ...      ...          ...   
2019-10-06  False  False          False   False  False    False        False   
2019-10-06  False  False          False   False  False    False        False   
2019-10-06  False  False          False   False  False    False        False   
2019-10-06  False  False          False   False  False    False        False   
2019-10-06  False  False          False 

↑ Data is highly right skewed (11.13). Max value is 497678.It is due to outliers. So, there has necessary to delete outliers. 

We use the **IQR method** to delete outliers from records

In [22]:
df_out = df[~((df < (Q1 - 1.5 * IQR)) |(df > (Q3 + 1.5 * IQR))).any(axis=1)]
print(df_out.shape)

(167132, 8)


In [23]:
df_out.reset_index().sort_values(by='Date')

Date    Year  Month  Day  Hour  Minute       ProductType  \
84394 2018-01-04  2018.0    1.0  4.0  17.0    50.0               Air   
84395 2018-01-04  2018.0    1.0  4.0  17.0    50.0               Air   
84375 2018-01-04  2018.0    1.0  4.0   8.0    26.0               Air   
84376 2018-01-04  2018.0    1.0  4.0   9.0    17.0               Air   
84378 2018-01-04  2018.0    1.0  4.0  12.0     0.0               Air   
...          ...     ...    ...  ...   ...     ...               ...   
63505 2019-12-05  2019.0   12.0  5.0  14.0    52.0               Air   
63506 2019-12-05  2019.0   12.0  5.0  14.0    57.0  Air Cancellation   
63507 2019-12-05  2019.0   12.0  5.0  14.0    58.0     Other Product   
63497 2019-12-05  2019.0   12.0  5.0  14.0    51.0               Air   
63583 2019-12-05  2019.0   12.0  5.0  17.0    49.0     Other Product   

      ItineraryType  NetFare  
84394      Domestic   3577.0  
84395      Domestic   3577.0  
84375      Domestic   8222.0  
84376      Domestic   3418.0  
84378      Domestic   3625.0  
...             ...      ...  
63505      Domestic   2305.0  
63506      Domestic  -3600.0  
63507      Domestic    127.0  
63497      Domestic   3150.0  
63583      Domestic    269.0  

[167132 rows x 9 columns]

Among CLEAN Data observations=**184800**, after delation of outlier REMAINDER Data observations=**167132** 

In [24]:
import warnings
warnings.filterwarnings('ignore')

### Create dummie variables of ['ProductType', 'ItineraryType']

In [25]:
dummy_df=pd.get_dummies(df_out, columns=['ProductType', 'ItineraryType'], drop_first=True)

In [26]:
dummy_df.rename(columns = {'ProductType_Air Cancellation':'ProTypeAirCanc','ProductType_Air Debit Note':'ProTypeAirDebtNote','ProductType_Air Loss':'ProTypeAirLoss','ProductType_Hotel':'ProTypeHot','ProductType_Hotel Cancellation':'ProTypeHotCanc','ProductType_Hotel Debit Note':'ProTypeHotDebNote','ProductType_Hotel Loss':'ProTypeHotLoss','ProductType_Other Product':'ProTypeOthPro','ProductType_Other Product Cancellation':'ProTypeOthProCanc','ProductType_Other Product Debit Note':'ProTypeOthProDebNote','ItineraryType_International':'IteTypeInter'}, inplace=True)


In [27]:
dummy_df.head(2)

Year  Month   Day  Hour  Minute  NetFare  ProTypeAirCanc  \
Date                                                                     
2018-01-12  2018.0    1.0  12.0  11.0    43.0  2160.00               0   
2018-01-12  2018.0    1.0  12.0  11.0    43.0     0.49               0   

            ProTypeAirDebtNote  ProTypeAirLoss  ProTypeHot  ProTypeHotCanc  \
Date                                                                         
2018-01-12                   0               0           0               0   
2018-01-12                   0               0           0               0   

            ProTypeHotDebNote  ProTypeHotLoss  ProTypeOthPro  \
Date                                                           
2018-01-12                  0               0              1   
2018-01-12                  0               0              1   

            ProTypeOthProCanc  ProTypeOthProDebNote  IteTypeInter  
Date                                                               
2018-01-12                  0                     0             0  
2018-01-12                  0                     0             0

In [28]:
dummy_df.dtypes

Year                    float64
Month                   float64
Day                     float64
Hour                    float64
Minute                  float64
NetFare                 float64
ProTypeAirCanc            uint8
ProTypeAirDebtNote        uint8
ProTypeAirLoss            uint8
ProTypeHot                uint8
ProTypeHotCanc            uint8
ProTypeHotDebNote         uint8
ProTypeHotLoss            uint8
ProTypeOthPro             uint8
ProTypeOthProCanc         uint8
ProTypeOthProDebNote      uint8
IteTypeInter              uint8
dtype: object

In [29]:
# converting dtypes using astype 
dummy_df["Year"]= dummy_df["Year"].astype(int) 
dummy_df["Month"]= dummy_df["Month"].astype(int) 
dummy_df["Day"]= dummy_df["Day"].astype(int) 
dummy_df["Hour"]= dummy_df["Hour"].astype(int) 
dummy_df["Minute"]= dummy_df["Minute"].astype(int) 
dummy_df["NetFare"]= dummy_df["NetFare"].astype(float) 
dummy_df["ProTypeHot"]= dummy_df["ProTypeHot"].astype(int) 
dummy_df["ProTypeOthPro"]= dummy_df["ProTypeOthPro"].astype(int)
dummy_df["IteTypeInter"]= dummy_df["IteTypeInter"].astype(int)

In [30]:
dummy_df.dtypes

Year                      int32
Month                     int32
Day                       int32
Hour                      int32
Minute                    int32
NetFare                 float64
ProTypeAirCanc            uint8
ProTypeAirDebtNote        uint8
ProTypeAirLoss            uint8
ProTypeHot                int32
ProTypeHotCanc            uint8
ProTypeHotDebNote         uint8
ProTypeHotLoss            uint8
ProTypeOthPro             int32
ProTypeOthProCanc         uint8
ProTypeOthProDebNote      uint8
IteTypeInter              int32
dtype: object

Year	Month	Day	Hour	Minute	NetFare	ProTypeAirCanc	ProTypeAirDebtNote	ProTypeAirLoss	ProTypeHot	ProTypeHotCanc	ProTypeHotDebNote	ProTypeHotLoss	ProTypeOthPro	ProTypeOthProCanc	ProTypeOthProDebNote	IteTypeInter

In [31]:
#data_model=dummy_df[['Year','Month','Day','Hour','Minute','ProTypeAirCanc','ProTypeAirDebtNote','ProTypeAirLoss','ProTypeHot','ProTypeHotCanc','ProTypeHotDebNote','ProTypeHotLoss','ProTypeOthPro','ProTypeOthProCanc','ProTypeOthProDebNote','IteTypeInter','NetFare']]

In [46]:
data_model=dummy_df[['Year','Month','Day','Hour','Minute','ProTypeHot','ProTypeOthPro','IteTypeInter','ProTypeAirCanc','NetFare']]

In [47]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score 

In [48]:
X = dummy_df[['Year','Month','Day','Hour','Minute','ProTypeHot','ProTypeOthPro','IteTypeInter','ProTypeAirCanc']]
y = dummy_df['NetFare']

In [49]:
# Split data into train and test 
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=0)

In [50]:
# Fit or Train the Model
lr_model = LinearRegression()
lr_model.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [51]:
score = lr_model.score(X_train,y_train)
print(f"Accuracy with train data : {score:0.2f}")

Accuracy with train data : 0.29


In [52]:
# Evaluate Model using test data 
y_pred =lr_model.predict(X_test)

In [53]:
# Find out accuracy with test data 
r2score = r2_score(y_test,y_pred)
print(f"Accuracy with test data :  {r2score:0.2f}")

Accuracy with test data :  0.28


In [54]:
# Pickle model
pd.to_pickle(lr_model,'30june_lr_model.pickle')

In [55]:
import pickle

In [56]:
with open('30june_lr_model.pickle', 'wb') as file:
    pickle.dump(lr_model, file)

In [57]:
with open('30june_lr_model.pickle', 'rb') as file:
    ml_reg=pickle.load(file)

In [60]:
# Take input from user
year = int(input("Enter Year    : "))
month = int(input("Enter Month  : "))
day = float(input("Enter Day : "))
hour = int(input("Enter Hour    : "))
minute = int(input("Enter Minute  : "))
protypehot = float(input("Enter ProTypeHot : "))
protypeothpro = int(input("Enter ProTypeOthPro    : "))
itetypeinter= int(input("Enter IteTypeInter  : "))
protypeaircanc=int(input("Enter ProTypeAirCanc : "))

Enter Year    : 2018
Enter Month  : 1
Enter Day : 12
Enter Hour    : 11
Enter Minute  : 43
Enter ProTypeHot : 0
Enter ProTypeOthPro    : 1
Enter IteTypeInter  : 0
Enter ProTypeAirCanc : 0


In [61]:
 ml_reg.predict([[year,month,day,hour,minute,protypehot,protypeothpro,itetypeinter,protypeaircanc]])

array([120.34249542])

2018	1	12	11	43	0	1	0

In [ ]:
# Take input from user
year = int(input("Enter Year    : "))
month = int(input("Enter Month  : "))
day = float(input("Enter Day : "))
hour = int(input("Enter Hour    : "))
minute = int(input("Enter Minute  : "))
protypehot = float(input("Enter ProTypeHot : "))
protypeothpro = int(input("Enter ProTypeOthPro    : "))
itetypeinter= int(input("Enter IteTypeInter  : "))

In [ ]:
ml_reg.predict([[year,month,day,hour,minute,protypehot,protypeothpro,itetypeinter]])

In [ ]:
import json,requests

In [ ]:
data_model[1:4]

url='http://localhost:9000/api'
data=json.dumps({'Year':2019,'Month':1,'Day':12,'Hour':11,'Minute':43,'ProTypeHot':0,'ProTypeOthPro':0,'IteTypeInter':1})
r=requests.post(url,data)
print (r.json())